File: api/internal/battle/battle.go
package battle

import (
	"context"
	"errors"
	"fmt"

	"github.com/ankeesler/spirits/api/internal/spirit"
)

const maxTurns = 100

func Run(ctx context.Context, spirits []*spirit.Spirit, onSpirits func([]*spirit.Spirit, error)) {
	turns := 0

	onSpirits(spirits, nil)

	s := newStrategy(spirits)
	for s.hasNext() {
		turns++
		if turns >= maxTurns {
			onSpirits(spirits, errors.New("too many turns"))
			return
		}
		from, to := s.next()

		if err := from.Action.Run(ctx, from, to); err != nil {
			onSpirits(spirits, fmt.Errorf("action errored: %w", err))
			return
		}

		onSpirits(spirits, nil)
	}
}

File: src/internal/battle.js
class Strategy {
  constructor(spirits) {
    this._spirits = spirits;
    this._next = 0;
  }

  hasNext() {
    return this._spirits.filter((spirit) => spirit.stats.health > 0).length > 1;
  }

  next() {
    return [this._spirits[this._next++ & 1], this._spirits[this._next & 1]];
  }
};

class Battle {
  constructor(spirits, callback) {
    this._spirits = spirits;
    this._callback = callback;
  }

  start() {
    this._callback(this._spirits);

    const strategy = new Strategy(this._spirits);
    while (strategy.hasNext()) {
      const [from, to] = strategy.next();
      from.action(from, to);

      this._callback(this._spirits);
    }
  }
};

module.exports = {
  Battle: Battle,
}